src/lib.rs

```diff
@@ -576,12 +670,13 @@ where
         for p in &mut projection {
             *p += 2;
         }
-        let mut fixed_projection = vec![0, 1, R::primary_key_index()];
+        let primary_key_index = self.schema.record_instance.primary_key_index::<R>();
+        let mut fixed_projection = vec![0, 1, primary_key_index];
         fixed_projection.append(&mut projection);
         fixed_projection.dedup();
 
         let mask = ProjectionMask::roots(
-            &arrow_to_parquet_schema(R::arrow_schema()).unwrap(),
+            &arrow_to_parquet_schema(&self.schema.record_instance.arrow_schema::<R>()).unwrap(),
```

Comment:
there is no need to call this method `self.schema.record_instance.arrow_schema::<R>()`, use `primary_key_index` directly